lib/executor/variable_arrival_rate.go

```diff
@@ -386,49 +300,59 @@ func (varr VariableArrivalRate) Run(ctx context.Context, out chan<- stats.Sample
 
 		return math.Min(1, float64(spent)/float64(duration)), right
 	}
+
 	varr.progress.Modify(pb.WithProgress(progresFn))
 	go trackProgress(ctx, maxDurationCtx, regDurationCtx, varr, progresFn)
 
 	regDurationDone := regDurationCtx.Done()
-	runIterationBasic := getIterationRunner(varr.executionState, varr.logger, out)
-	runIteration := func(vu lib.VU) {
-		runIterationBasic(maxDurationCtx, vu)
-		vus <- vu
-	}
+	runIteration := getIterationRunner(varr.executionState, varr.logger, out)
 
 	remainingUnplannedVUs := maxVUs - preAllocatedVUs
-	rateChangesStream := varr.streamRateChanges(maxDurationCtx, startTime)
 
-	for {
+	var timer = time.NewTimer(time.Hour)
+	var start = time.Now()
+	var ch = make(chan time.Duration, 10) // buffer 10 iteration times ahead
+	var prevTime time.Duration
+	go varr.config.cal(varr.executionState.ExecutionTuple, ch)
```

Comment:
This seems like an excellent thing to put in the `Init()` method. The only problem that I can see is that this goroutine is going to leak. Not a big deal, but still not really great, given that test runs can have multiple executors, some of which working only at the start

Reply:
I would prefer to leave it for when I rewrite the calculation as a state machine - maybe then it will be a bigger problem.

I don't think this takes enough time to warrant be moved and adding `ch` as a field to the executor's struct.

Do you mean that it leaks until it finishes ? is needed? because - yes that it does. But it will end when it goes through all the stages and emits all the iterations needed